pkg/iter/iterator_test.go

```diff
@@ -231,50 +230,67 @@ func inverse(g generator) generator {
-	// First is most common.
-	tuples := []tuple{
-		{Entry: logproto.Entry{Line: "a"}},
-		{Entry: logproto.Entry{Line: "b"}},
-		{Entry: logproto.Entry{Line: "c"}},
-		{Entry: logproto.Entry{Line: "a"}},
-		{Entry: logproto.Entry{Line: "b"}},
-		{Entry: logproto.Entry{Line: "c"}},
-		{Entry: logproto.Entry{Line: "a"}},
-	}
-	require.Equal(t, "a", mostCommon(tuples).Entry.Line)
-
-	tuples = []tuple{
-		{Entry: logproto.Entry{Line: "a"}},
-		{Entry: logproto.Entry{Line: "b"}},
-		{Entry: logproto.Entry{Line: "b"}},
-		{Entry: logproto.Entry{Line: "c"}},
-		{Entry: logproto.Entry{Line: "c"}},
-		{Entry: logproto.Entry{Line: "c"}},
-		{Entry: logproto.Entry{Line: "d"}},
-	}
-	require.Equal(t, "c", mostCommon(tuples).Entry.Line)
-}
-
-func TestInsert(t *testing.T) {
-	toInsert := []tuple{
-		{Entry: logproto.Entry{Line: "a"}},
-		{Entry: logproto.Entry{Line: "e"}},
-		{Entry: logproto.Entry{Line: "c"}},
-		{Entry: logproto.Entry{Line: "b"}},
-		{Entry: logproto.Entry{Line: "d"}},
-		{Entry: logproto.Entry{Line: "a"}},
-		{Entry: logproto.Entry{Line: "c"}},
+func TestHeapIteratorDeduplication(t *testing.T) {
+	foo := logproto.Stream{
+		Labels: `{app="foo"}`,
+		Entries: []logproto.Entry{
+			{Timestamp: time.Unix(0, 1), Line: "1"},
+			{Timestamp: time.Unix(0, 2), Line: "2"},
+			{Timestamp: time.Unix(0, 3), Line: "3"},
+		},
 	}
-
-	var ts []tuple
-	for _, e := range toInsert {
-		ts = insert(ts, e)
+	bar := logproto.Stream{
+		Labels: `{app="bar"}`,
+		Entries: []logproto.Entry{
+			{Timestamp: time.Unix(0, 1), Line: "1"},
+			{Timestamp: time.Unix(0, 2), Line: "2"},
+			{Timestamp: time.Unix(0, 3), Line: "3"},
+		},
 	}
+	assertIt := func(it EntryIterator) {
+		for i := 0; i < 3; i++ {
+			require.True(t, it.Next())
+			require.NoError(t, it.Error())
+			require.Equal(t, bar.Labels, it.Labels())
+			require.Equal(t, bar.Entries[i], it.Entry())
 
-	require.True(t, sort.SliceIsSorted(ts, func(i, j int) bool {
-		return ts[i].Line < ts[j].Line
-	}))
+			require.True(t, it.Next())
+			require.NoError(t, it.Error())
+			require.Equal(t, foo.Labels, it.Labels())
+			require.Equal(t, foo.Entries[i], it.Entry())
+
+		}
+		require.False(t, it.Next())
+		require.NoError(t, it.Error())
+	}
+	// forward iteration
+	it := NewHeapIterator(context.Background(), []EntryIterator{
+		NewStreamIterator(foo),
+		NewStreamIterator(bar),
+		NewStreamIterator(foo),
+		NewStreamIterator(bar),
+		NewStreamIterator(foo),
+		NewStreamIterator(bar),
+		NewStreamIterator(foo),
+	}, logproto.FORWARD)
+	assertIt(it)
+
+	// backward iteration
+	it = NewHeapIterator(context.Background(), []EntryIterator{
+		NewStreamIterator(foo),
+		NewStreamIterator(bar),
+		NewStreamIterator(foo),
+		NewStreamIterator(bar),
+		NewStreamIterator(foo),
+		NewStreamIterator(bar),
+		NewStreamIterator(foo),
+	}, logproto.BACKWARD)
+	// first reverse streams, they should already be correctly ordered for the heap iterator to work.
+	for i, j := 0, len(foo.Entries)-1; i < j; i, j = i+1, j-1 {
```

Comment:
This is actually mutating the internal entries used by the stream iter as they reference the same slice. I think you'll want a function which creates the same `[]stream` to guarantee we don't mutate the same underlying data. Basically the testware and the HeapIter should use identical but separate copies of the same data.